File: 1001-2000/1856-maximum-subarray-min-product/1856-maximum-subarray-min-product.cpp
class Solution {
public:
    int maxSumMinProduct(vector<int>& nums) {
        int size = nums.size();
        long long mod = 1e9 + 7;

        vector<long long> prefix(size+1, 0);
        for(int i=0;i<size;i++)
        {
            prefix[i+1] = prefix[i] + nums[i];
        }

        vector<int> st;
        long long max_product = 0;
        for(int i=0;i<=size;i++)
        {
            long long curr_val = (i==size) ? 0 : nums[i];
            while(!st.empty() && nums[st.back()] > curr_val)
            {
                long long min_val = nums[st.back()];
                st.pop_back();

                int left_bound = st.empty() ? -1 : st.back();

                long long subarray_sum = prefix[i] - prefix[left_bound + 1];

                max_product = max(max_product, min_val*subarray_sum);
            }
            st.push_back(i);
        }
        return max_product % mod;
    }
};

File: 1001-2000/1856-maximum-subarray-min-product/1856-maximum-subarray-min-product.java
class Solution {
    public int maxSumMinProduct(int[] nums) {

        int mod = 1_000_000_007 ;

        int n = nums.length ;

        long[] preSum = new long[n + 1] ;
        for(int i = 0 ; i < n ; i ++)
        {
            preSum[i + 1] = preSum[i] + nums[i] ;
        }

        int[] stack = new int[n + 1] ;
        int top = -1 ;

        long res = 0 ;

        for(int i = 0 ; i <= n ; i ++)
        {
            while(top >= 0 && (i == n || nums[stack[top]] > nums[i]))
            {
                int min = nums[stack[top --]] ;

                int prevIdx = top < 0 ? 0 : stack[top] + 1 ;

                res = Math.max(res, 1L * min * (preSum[i] - preSum[prevIdx])) ;
            }
            stack[++ top] = i ;

        }
        return (int)(res % mod) ;
    }
}        